rust/src/measurements/permute_and_flip/mod.rs

```diff
@@ -0,0 +1,123 @@
+    core::{Function, Measurement, PrivacyMap},
+    domains::{AtomDomain, VectorDomain},
+    error::Fallible,
+    measurements::Optimize,
+    measures::MaxDivergence,
+    metrics::LInfDistance,
+    traits::samplers::{sample_bernoulli_exp, CastInternalRational, SampleUniformIntBelow},
+    traits::{DistanceConstant, Float, Number},
+};
+use dashu::base::Sign;
+use dashu::rational::RBig;
+
+fn exact_fisher_yates(n: usize, trials_per_coin: Option<usize>) -> Fallible<Vec<usize>> {
+    let mut permutation: Vec<usize> = (0..n).collect();
+
+    for i in (1..n).rev() {
+        let j = usize::sample_uniform_int_below(i, trials_per_coin)?;
+        permutation.swap(i, j);
+    }
+
+    Ok(permutation)
+}
+
+/// Make a Measurement that takes a vector of scores and privately selects the index of the highest score via permute and flip.
+///
+/// # Arguments
+/// * `input_domain` - Domain of the input vector. Must be a non-nullable VectorDomain.
+/// * `input_metric` - Metric on the input domain. Must be LInfDistance
+/// * `scale` - Higher scales are more private.
+/// * `optimize` - Indicate whether to privately return the "Max" or "Min"
+///
+/// # Generics
+/// * `TIA` - Atom Input Type. Type of each element in the score vector.
+/// * `QO` - Output Distance Type.
+pub fn make_report_noisy_max_permute_and_flip<TIA, QO>(
+    input_domain: VectorDomain<AtomDomain<TIA>>,
+    input_metric: LInfDistance<TIA>,
+    scale: QO,
+    optimize: Optimize,
+) -> Fallible<Measurement<VectorDomain<AtomDomain<TIA>>, usize, LInfDistance<TIA>, MaxDivergence<QO>>>
+where
+    TIA: Number + CastInternalRational,
+    QO: CastInternalRational + DistanceConstant<TIA> + Float,
+{
+    if input_domain.element_domain.nullable() {
+        return fallible!(MakeMeasurement, "input domain must be non-nullable");
+    }
+
+    if scale.is_sign_negative() {
+        return fallible!(MakeMeasurement, "scale must not be negative");
+    }
+
+    let scale_frac: RBig = scale.into_rational()?;
+    let sgn: Sign = match optimize {
+        Optimize::Max => Sign::Positive,
+        Optimize::Min => Sign::Negative,
+    };
+
+    Measurement::new(
+        input_domain,
+        Function::new_fallible(move |arg: &Vec<TIA>| {
+            let permutation = exact_fisher_yates(arg.len(), None)?;
+            let rational_elements = arg
+                .iter()
+                .map(|x| ((*x).into_rational().map(|x| x * sgn)))
+                .collect::<Fallible<Vec<RBig>>>()?;
+
+            // get argmax of the rational elements
+            let argmax = rational_elements
+                .iter()
+                .enumerate()
+                .max_by_key(|&(_, value)| value)
+                .map(|(index, _)| index)
+                .ok_or(err!(FailedFunction, "there must be at least one candidate"))?;
+
+            let max_element = rational_elements
+                .get(argmax)
+                .ok_or(err!(FailedFunction, "there must be at least one candidate"))?;
```

Comment:
I don't think you ever need the index of the max score, so the map could just return the value.